HW2/P3/mandelbrot.pyx

```diff
@@ -0,0 +1,148 @@
+import numpy as np
+cimport numpy as np
+cimport cython
+import numpy
+cimport AVX
+from cython.parallel import prange
+
+#We adapt the magnitude_squared function for an AVX.float8 element (we follow magnitude_squared = z_real^2 + z_imag^2)
+cdef AVX.float8 magnitude_squared(AVX.float8 z_real, AVX.float8 z_imag) nogil:
+    return AVX.fmadd(z_real, z_real, AVX.mul(z_imag, z_imag))
+
+
+#We create a function to copy the temporary tmp_counts buffer to the out_counts array. 
+cdef void counts_to_output(AVX.float8 counts,
+                      np.uint32_t [:, :] out_counts,
+                      int i, int j) nogil:
+    
+    cdef:
+        float tmp_counts[8]
+        int jj
+        
+    AVX.to_mem(counts, &(tmp_counts[0])) 
+    
+    for jj in range(8):
+        out_counts[i, 8*j+jj] =  <int>tmp_counts[jj]
+         
+                
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+                 np.uint32_t [:, :] out_counts,
+                 int max_iterations=511):
+    cdef:
+       int i, j, iter
+       AVX.float8 c_real, c_imag, z_real, z_imag, mask, accepted, iterations, ones, z_real_tmp
+
+       # To declare AVX.float8 variables, use:
+       # cdef:
+       #     AVX.float8 v1, v2, v3
+       #
+       # And then, for example, to multiply them
+       #     v3 = AVX.mul(v1, v2)
+       #
+       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+
+    assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
+    assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+
+    with nogil:
+       
+        for i in prange(in_coords.shape[0], schedule='static',chunksize=1,num_threads = 4):
+            for j in range(in_coords.shape[1]/8):
```

Comment:
Clever way of iterating through all of the coordinates in steps of 8. Nicely done